base/hashing.jl

```diff
@@ -108,12 +86,87 @@ else
     hash(x::QuoteNode, h::UInt) = hash(x.value, h + 0x469d72af)
 end
 
-## hashing strings ##
+hash(x::Symbol) = objectid(x)
+
 
-const memhash = UInt === UInt64 ? :memhash_seed : :memhash32_seed
-const memhash_seed = UInt === UInt64 ? 0x71e729fd56419c81 : 0x56419c81
+load_le(::Type{T}, ptr::Ptr{UInt8}, i) where {T <: Union{UInt32, UInt64}} =
+    unsafe_load(convert(Ptr{T}, ptr + i - 1))
 
-@assume_effects :total function hash(s::String, h::UInt)
-    h += memhash_seed
-    ccall(memhash, UInt, (Ptr{UInt8}, Csize_t, UInt32), s, sizeof(s), h % UInt32) + h
+function read_small(ptr::Ptr{UInt8}, n::Int)
+    return (UInt64(unsafe_load(ptr)) << 56) |
+        (UInt64(unsafe_load(ptr, div(n + 1, 2))) << 32) |
+        UInt64(unsafe_load(ptr, n))
 end
+
+function hash_bytes(
```

Comment:
This function has most effect bits - except `terminates_locally` and `noub` as far as I can tell. I don't know what effects are inferred, but maybe you could add any missing effects.

Reply:
it's definitely `terminates_locally` but iiuc `!noub` is correct since I'm doing an `unsafe_load` on pointers that have to be validated by the caller

Reply:
`nothrow` I expect would be fixed with the resolution of https://github.com/JuliaLang/julia/issues/57557